### PR TITLE
[SPARK-40414][SQL][PYTHON] More generic type on PythonArrowInput and PythonArrowOutput

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to change PythonArrowInput and PythonArrowOutput to be more generic to cover the complex data type on both input and output. This is a baseline work for #37863.

### Why are the changes needed?

The traits PythonArrowInput and PythonArrowOutput can be further generalized to cover complex data type on both input and output. E.g. Not all operators would have simple InternalRow as input data to pass to Python worker and vice versa for output data.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @HyukjinKwon @ueshin Please take a look, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala [hunk 1]
```diff
@@ -33,12 +33,14 @@ import org.apache.spark.sql.vectorized.{ArrowColumnVector, ColumnarBatch, Column
 
 /**
  * A trait that can be mixed-in with [[BasePythonRunner]]. It implements the logic from
- * Python (Arrow) to JVM (ColumnarBatch).
+ * Python (Arrow) to JVM (output type being deserialized from ColumnarBatch).
  */
-private[python] trait PythonArrowOutput { self: BasePythonRunner[_, ColumnarBatch] =>
+private[python] trait PythonArrowOutput[OUT <: AnyRef] { self: BasePythonRunner[_, OUT] =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala:38]: qq: should it be `<: AnyRef`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala [hunk 2]
```diff
@@ -26,21 +26,30 @@ import org.apache.spark.{SparkEnv, TaskContext}
...
 import org.apache.spark.util.Utils
 
 /**
  * A trait that can be mixed-in with [[BasePythonRunner]]. It implements the logic from
- * JVM (an iterator of internal rows) to Python (Arrow).
+ * JVM (an iterator of internal rows + additional data if required) to Python (Arrow).
  */
-private[python] trait PythonArrowInput { self: BasePythonRunner[Iterator[InternalRow], _] =>
+private[python] trait PythonArrowInput[IN] { self: BasePythonRunner[IN, _] =>
+  protected val sqlConf = SQLConf.get
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala:39]: Seems like we don't need this (in https://github.com/apache/spark/pull/37863, it's not used too)

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala [hunk 3]
```diff
@@ -107,3 +107,27 @@ private[python] trait PythonArrowInput { self: BasePythonRunner[Iterator[Interna
...
 }
+
+private[python] trait BasicPythonArrowInput extends PythonArrowInput[Iterator[InternalRow]] {
+  self: BasePythonRunner[Iterator[InternalRow], _] =>
+
+  protected def writeIteratorToArrowStream(
+    root: VectorSchemaRoot,
+    writer: ArrowStreamWriter,
+    dataOut: DataOutputStream,
+    inputIterator: Iterator[Iterator[InternalRow]]): Unit = {
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala:118]: nit: indent?

**Review by ueshin — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowInput.scala:39]: Ah OK didn't notice this. Will remove this. Thanks for the pointer.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/PythonArrowOutput.scala:38]: We assign `null` to the OUT type (although that's a trick) hence need to be AnyRef at least if I understand correctly.

**Comment by HeartSaVioR** [comment]
I'll merge this PR once build is green.

**Comment by HeartSaVioR** [comment]
Thanks for reviewing! Merging to master.

**Pull request closed without merging**